culane_metric/evaluate.py

```diff
@@ -0,0 +1,205 @@
+
+This script will compute the F1, precision and recall metrics as described in the CULane benchmark.
+
+The predictions format is the same one used in the CULane benchmark.
+In summary, for every annotation file:
+    labels/a/b/c.json
+There should be a prediction file:
+    predictions/a/b/c.lines.txt
+Inside each .lines.txt file each line will contain a sequence of points (x, y) separated by spaces.
+For more information, please see https://xingangpan.github.io/projects/CULane.html
+
+This script uses two methods to compute the IoU: one using an image to draw the lanes (named `discrete` here) and
+another one that uses shapes with the shapely library (named `continuous` here). The results achieved with the first
+method are very close to the official CULane implementation. Although the second should be a more exact method and is
+faster to compute, it deviates more from the official implementation. By default, the method closer to the official
+metric is used.
+"""
+
+import os
+import argparse
+from functools import partial
+
+import cv2
+import numpy as np
+from p_tqdm import t_map, p_map
+from scipy.interpolate import splprep, splev
+from scipy.optimize import linear_sum_assignment
+from shapely.geometry import LineString, Polygon
+
+import unsupervised_llamas.common.helper_scripts as helper_scripts
+import unsupervised_llamas.label_scripts.spline_creator as spline_creator
+
+LLAMAS_IMG_RES = (717, 1276)
+
+
+def add_ys(xs, vertical_cutoff=300):
+    assert len(xs) >= 717 - vertical_cutoff, "Need at least 417 pixels per lane"
+
+    # Reference lanes go from 0 to 717. If a horizontal entry is not
+    # defined, it is stored as -1. We have to filter for that.
+
+    if len(xs) == 717:  # lane regressed across complete image
+        xs = xs[vertical_cutoff:]
+    elif len(xs) == 417:  # lane regress across part of image that is relevant
+        pass
+    else:
+        raise NotImplementedError(f"Evaluations not implemented for length of detected lane: {len(xs)}")
+    xs = np.array(xs)
+    valid = xs >= 0
+    xs = xs[valid]
+    assert len(xs) > 1
+    ys = np.arange(300, 717)[valid]
+    return list(zip(xs, ys))
+
+
+def draw_lane(lane, img=None, img_shape=None, width=30):
+    if img is None:
+        img = np.zeros(img_shape, dtype=np.uint8)
+    lane = lane.astype(np.int32)
+    for p1, p2 in zip(lane[:-1], lane[1:]):
+        cv2.line(img, tuple(p1), tuple(p2), color=(1,), thickness=width)
+    return img
+
+
+def discrete_cross_iou(xs, ys, width=30, img_shape=LLAMAS_IMG_RES):
+    xs = [draw_lane(lane, img_shape=img_shape, width=width) > 0 for lane in xs]
+    ys = [draw_lane(lane, img_shape=img_shape, width=width) > 0 for lane in ys]
+
+    ious = np.zeros((len(xs), len(ys)))
+    for i, x in enumerate(xs):
+        for j, y in enumerate(ys):
+            ious[i, j] = (x & y).sum() / (x | y).sum()
+    return ious
+
+
+def continuous_cross_iou(xs, ys, width=30, img_shape=LLAMAS_IMG_RES):
+    h, w, _ = img_shape
```

Comment:
from common/constants.py

```suggestion
    h, w = constants.IMAGE_HEIGHT, constants.IMAGE_WIDTH
```